q2cli/cli.py

```diff
@@ -80,22 +91,66 @@ def _echo_info(ctx, name, value):
 def cli():
     pass
 
-# TODO: update keys to be the types (rather than their str representations)
-# pending biocore/qiime2#46
-#_type_map = {int: click.INT, str: click.STRING, float: click.FLOAT}
-
 
-def _create_callback(wf):
+def _build_workflow_callback(wf):
     def f(ctx, **kwargs):
         # execute workflow
         executor = SubprocessExecutor()
+        # TODO remove hardcoding of extension pending
+        # https://github.com/qiime2/qiime2/issues/59
+        output_extension = '.qza'
         inputs = {
             ia_name: kwargs[ia_name] for ia_name in wf.signature.inputs}
         parameters = {
             ip_name: kwargs[ip_name] for ip_name in wf.signature.parameters}
-        outputs = {
-            oa_name: kwargs[oa_name] for oa_name in wf.signature.outputs
-        }
+        outputs = {}
+        for oa_name in wf.signature.outputs:
+            oa_value = kwargs[oa_name]
+            file_extension = os.path.splitext(oa_value)[1]
+            if file_extension != output_extension:
+                oa_value = ''.join([oa_value, output_extension])
+            outputs[oa_name] = oa_value
+        future_, _ = executor(wf, inputs, parameters, outputs)
+
+        # block (i.e., wait) until result is ready
+        completed_process = future_.result()
+
+        if completed_process.returncode != 0:
+            click.echo(completed_process.stdout)
+            click.echo(completed_process.stderr, err=True)
+            ctx.exit(completed_process.returncode)
+
+    return click.pass_context(f)
+
+
+def _build_visualization_callback(wf):
+    # TODO there is a lot of code duplicated between
+    # _build_visualization_callback and _build_workflow_callback - revisit
+    # this after the refactoring that is happening as part of
+    # https://github.com/qiime2/qiime2/issues/39
+    def f(ctx, **kwargs):
+        # TODO nest output_dir under user-configured temporary directory,
+        # pending https://github.com/qiime2/qiime2/issues/12
+        output_dir = tempfile.TemporaryDirectory()
```

Comment:
Ideally this will be provided by the framework as well, so that the user using the `Artifact` API will just provide the "actual" arguments, and the SDK will fill in the temporary directory implicitly (managed by a configuration)